src/operations/blas2/gemv.hpp

```diff
@@ -80,6 +80,261 @@ SYCL_BLAS_INLINE void AddSetColumns<rhs_t>::adjust_access_displacement() {
+    value_t *matrix_scratch = local_mem.localAcc.get_pointer() + local_range;
+
+    for (index_t c_tile_id = 0; c_tile_id < tile_c_loops; ++c_tile_id) {
+      // Extract a matrix block from global memory
+      extract_input_block(matrix_scratch, local_id, group_id, local_range, lda,
+                          c_tile_id);
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+      index_t mat_index = local_id;
+
+#pragma unroll
+      for (index_t c_dim_id = 0; c_dim_id < cl_elems; ++c_dim_id) {
+        sum = cl::sycl::mad(matrix_scratch[mat_index], *vector_scratch++, sum);
+
+        mat_index += local_range + 1;  // Adding one as bank offset
+      }
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+    }
+
+    const index_t out_index = nc_dim_index + (c_group_id * nc_dim);
+
+    if (nc_dim_index < nc_dim) {
+      lhs_.eval(out_index) = sum;
+    }
+    return sum;
+  }
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          bool is_transposed, int cache_line_size, int work_per_thread>
+template <typename ScratchPointerType>
+SYCL_BLAS_INLINE void
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size,
+     work_per_thread>::extract_input_block(ScratchPointerType matrix_scratch,
+                                           const index_t &local_id,
+                                           const index_t &group_id,
+                                           const index_t &local_range,
+                                           const index_t &lda,
+                                           index_t c_tile_id) {
+  constexpr int cl_elems = cache_line_size / sizeof(value_t);
+
+  // Group indices within global grid
+  const index_t nc_group_id = group_id / WGs_per_C_;
+  const index_t c_group_id = group_id - nc_group_id * WGs_per_C_;
+
+  // Tile dimensions
+  const int tile_dim_nc = local_range / cl_elems;
+  constexpr int tile_dim_c = cl_elems;
+
+  // In-tile indices
+  const index_t tile_nc_index = local_id / tile_dim_c;
+  const index_t tile_c_index = local_id % tile_dim_c;
+
+  // Tile-group (local) indices
+  index_t local_nc_index = tile_nc_index;
+  const index_t local_c_index = tile_c_index + c_tile_id * tile_dim_c;
+
+  // Global (grid) indices
+  const index_t grid_nc_index = local_nc_index + nc_group_id * local_range;
+  const index_t grid_c_index = local_c_index + c_group_id * local_range;
+
+  // Matrix index calculated based on global grid indices
+  index_t mat_index = grid_c_index + grid_nc_index * lda;
+
+  // Number of iterations to load all tiles
+  // local_range / tile_dim_nc == cl_elems
+  constexpr int tile_nc_loops = cl_elems;
+
+  // Transposed index in scratch-space with bank offset which adds 1 word of
+  // padding every 'tile_dim_c' elements
+  const index_t scratch_index = tile_c_index * local_range + tile_c_index;
```

Comment:
Can you write that `scratch_index = (tile_c_index + 1) * local_range`?

Reply:
Yup. `tile_c_index` had an alias before for the bank offset, hence the mistake here. (compiler would/should optimize the old code either way).

Reply:
Actually it's `scratch_index = tile_c_index * (local_range + 1)` sorry